Add `DimensionControl` component

getdave opened the pull request:
![Screen Shot 2019-07-29 at 13 59 53](https://user-images.githubusercontent.com/444434/62050467-2cf1ba80-b209-11e9-91b3-db7f5da37b80.png)


Sub PR of [Adds basic dimension controls to Group Block](https://github.com/WordPress/gutenberg/pull/16730).

Adds a standardised `<DimensionControl>` component which provides a UI for changing dimension/spacing values.

* Standardised, accessible markup
* Customisable label
* Optional customisable icon

## Please note

This isn't in any way wired up to Block state so changing the values doesn't do anything. This is purely a PR for the UI only. 

## Testing Instructions

To manually test this component you will need to add it to the `core/group` Block. @jorgefilipecosta has prepared a patch for this. After checking out this PR, save and then apply the following patch using `git apply %%saved-name-of-patch%%`:

```
diff --git a/packages/block-library/src/group/edit.js b/packages/block-library/src/group/edit.js
index 1e7445a20..c22580ad5 100644
--- a/packages/block-library/src/group/edit.js
+++ b/packages/block-library/src/group/edit.js
@@ -2,10 +2,12 @@
  * External dependencies
  */
 import classnames from 'classnames';
+import { partialRight } from 'lodash';
 
 /**
  * WordPress dependencies
  */
+import { useState } from '@wordpress/element';
 import { withSelect } from '@wordpress/data';
 import { compose } from '@wordpress/compose';
 import { __ } from '@wordpress/i18n';
@@ -16,6 +18,35 @@ import {
 	withColors,
 } from '@wordpress/block-editor';
 
+import { DimensionControl } from '@wordpress/components';
+
+function TestDimensionControl() {
+	const [ dimensions, setDimensions ] = useState( {} );
+	const { paddingSize } = dimensions;
+
+	const updateSpacing = ( dimension, size, device = '' ) => {
+		setDimensions( {
+			...dimensions,
+			[ `${ dimension }${ device }` ]: size,
+		} );
+	};
+	const resetSpacingDimension = ( dimension, device = '' ) => {
+		setDimensions( {
+			...dimensions,
+			[ `${ dimension }${ device }` ]: '',
+		} );
+	};
+	return (
+		<DimensionControl
+			label={ __( 'Padding' ) }
+			icon={ 'desktop' }
+			onReset={ partialRight( resetSpacingDimension, 'paddingSize' ) }
+			onChange={ partialRight( updateSpacing, 'paddingSize' ) }
+			currentSize={ paddingSize }
+		/>
+	);
+}
+
 function GroupEdit( {
 	className,
 	setBackgroundColor,
@@ -43,6 +74,7 @@ function GroupEdit( {
 						},
 					] }
 				/>
+				<TestDimensionControl />
 			</InspectorControls>
 			<div className={ classes } style={ styles }>
 				<div className="wp-block-group__inner-container">


```


## How has this been tested?

I've temporarily hacked and implementation into the `core/group` Block which utilises `<select>` inputs and temporary `state` for demo purposes. 

Note that **this will be removed before merging**.


## Screenshots

See above. 

#### Toggle open
<img width="280" alt="Screen Shot 2019-07-29 at 13 53 11" src="https://user-images.githubusercontent.com/444434/62050076-7b528980-b208-11e9-8376-3ea630f04cd6.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

getdave committed:
Adds initial component

Note this is copied wholescale from original PR https://github.com/WordPress/gutenberg/pull/16730

getdave committed:
Remove redunant files. Refactors tests.

getdave committed:
Updates docs

getdave committed:
Checks callbacks are functions prior to calling

getdave committed:
Adds temp testing example usage of component to Group Block

getdave committed:
Updates to allow sizes as an (optionaly) prop dependency

karmatosed commented:
Just going to pop a little design review here, thanks for this.

![image](https://user-images.githubusercontent.com/253067/62366213-338c7480-b51e-11e9-832d-aad62d2605ca.png)

- The section being called 'spacing' but seeing 'padding' threw me a little.
- I see here small through to extra-large but it was hard to work out 'what those values' are.
- I didn't seem to be able to get any changes to the padding despite saving and viewing a few times. It also seemed to get stuck on reload back to default option. This was in Chrome.

shaunandrews commented:
My initial thought here was to try using a [Range component](https://developer.wordpress.org/block-editor/components/range-control/), which came out like this:

<img width="642" alt="image" src="https://user-images.githubusercontent.com/191598/62391930-19f62780-b534-11e9-87b2-1d82ed96f980.png">

However, when considering #16790 and other scenarios where we might have multiple controls listed the Range component gets really messy and distracting.

<img width="649" alt="image" src="https://user-images.githubusercontent.com/191598/62391999-44e07b80-b534-11e9-9e00-6237f01725d8.png">

--

So I tried sticking with the dropdown but removing the Icon and Label elements:

<img width="627" alt="image" src="https://user-images.githubusercontent.com/191598/62392085-82dd9f80-b534-11e9-85ac-6313c3c8e46b.png">

This is simpler overall, and lends itself better to showing multiple controls in a row.

--

I notice you have a default value of "Please select...". I find this placeholder value confusing. Instead, perhaps we add a "Default" value:

<img width="418" alt="image" src="https://user-images.githubusercontent.com/191598/62392674-12378280-b536-11e9-835c-b2ed30dcddbb.png">

kjellr commented:
Thanks for breaking this out into a smaller issue, @getdave. 

A base concern I have about this change is: How can we make this manageable for themes? On a basic level, we'll want themes to be able to specify what exactly these sizes mean. (In the near term at least). Similar to how [themes can provide custom font sizes](https://developer.wordpress.org/block-editor/developers/themes/theme-support/#block-font-sizes). 

I'm relatively certain that the main point of difficulty from theme development is the fact that any margins that are added here will need to be overwritten by themes for any full-width innerblocks. We already have a little taste of what this is like based on the simple `30px` padding added by default to Group blocks with a background color. The Gutenberg Starter theme manually adjusts the width, max-width, and position of full-width inner blocks to override that: 

https://github.com/WordPress/gutenberg-starter-theme/blob/043fb28c64ee370551d4df1ac3611c4b4385604e/css/blocks.css#L136-L141

Twenty Nineteen handles this similarly. It [rewrites that `30px` to align its base styles](https://core.trac.wordpress.org/browser/trunk/src/wp-content/themes/twentynineteen/sass/blocks/_blocks.scss#L820), and then it [adjusts the width, max-width, and margin](https://core.trac.wordpress.org/browser/trunk/src/wp-content/themes/twentynineteen/sass/blocks/_blocks.scss#L854) of full-width blocks inside it to match. 

To get this working with custom sizes, themes will need to essentially write in that same change, but for all supported custom spacing sizes.

getdave commented:
@shaunandrews Thanks for your feedback and insight. It is much appreciated.

> sticking with the dropdown but removing the Icon and Label elements:

I like the approach of removing the icon for the simple controls (ie: when you don't have ["responsive" controls](https://github.com/WordPress/gutenberg/pull/16790). I think the icon is probably necessary for the responsive mode but `icon` can be an optional prop on the component so we're all good.

From an a11y perspective removing a visual label could be problematic. I also feel it might affect UX. The mockup you seem to have landed on still shows a label which suggests you also came to this conclusion. Can you confirm?

> I notice you have a default value of "Please select...". I find this placeholder value confusing. Instead, perhaps we add a "Default" value:

You are right. `Please select...` isn't clear that it is basically saying "revert to default setting". I'll try your suggestion.

getdave commented:
@kjellr Thanks for your insightful feedback on Theme issues.

> Thanks for breaking this out into a smaller issue, @getdave.

No problem!

> On a basic level, we'll want themes to be able to specify what exactly these sizes mean. (In the near term at least). Similar to how themes can provide custom font sizes.

I agree. I was hunting in the docs to see if this was already a thing but it isn't and I think it should be. How can we make this happen? What's the impact for Themes? Can you help?

> I'm relatively certain that the main point of difficulty from theme development is the fact that any margins that are added here will need to be overwritten by themes for any full-width innerblocks. 

Yes that's true. I remember what a minefield writing that CSS was (if you recall)!

I'm keen to discover what the true scale of problem is  we looking at? Is it huge or manageable?

From my understanding, our options seem to be the following:


1. We add some logic (external to the component itself as this is not the concern of the component itself) which disables the UI for (and impact of) `margin` for Group blocks which are set to full-width alignment.

2. We add support to the Gutenberg and Twenty Nineteen themes to support the S, M, L, XL sizes we define.

3. We produce docs advising Theme devs how to make the necessary CSS changes.

My questions are:

- which of the above options are _desirable_?
- which of the above options are _feasible_?
- which of the above options are _required_ for this PR to land?
- are any of the above options insurmountable (if so then I'd rather can this right now 😄 )?

Much appreciated

kjellr commented:
```
which of the above options are desirable?
which of the above options are feasible?
which of the above options are required for this PR to land?
are any of the above options insurmountable (if so then I'd rather can this right now 😄 )?
```

Good questions! I definitely don't think any of this is insurmountable! So that's the good news. 😄 I'll let others speak to the feasibility, but I can try to outline what I think are the desirable + required 
options. 

## Desirable

I _think_ the ideal state would be: 

1. Similar to wide/full images, we offer a theme opt-in for custom margins in general. If this is not enabled, these controls don't show up. That way there's basically no chance this will break any theme. 
2. We offer a theme opt-in for themes to provide their own custom sizes. This is the equivalent of the [custom font size](https://developer.wordpress.org/block-editor/developers/themes/theme-support/#block-font-sizes) options Gutenberg already provides. 
3. We update the default themes to support both of those.
4. We write up documentation for theme developers to adopt both. 

We could theoretically make 1 & 2 variations of the same opt-in. So a basic opt-in would be: 

```
add_theme_support( 'editor-margin-sizes' );
```
or, you could add an optional array to specify custom sizes: 
```
add_theme_support( 'editor-margin-sizes', array(
    array(
        'name' => __( 'Small', 'themeLangDomain' ),
        'size' => 8px,
        'slug' => 'small'
    ),
    array(
        'name' => __( 'Normal', 'themeLangDomain' ),
        'size' => 16,
        'slug' => 'medium'
    ),
    array(
        'name' => __( 'Large', 'themeLangDomain' ),
        'size' => 24,
        'slug' => 'large'
    ),
    array(
        'name' => __( 'Huge', 'themeLangDomain' ),
        'size' => 36,
        'slug' => 'extra-large'
    )
) );
```

Themes are opted into custom font sizes whether or not they declare theme_support, so this works a little differently... but it's cleaner than forcing themes to do two opt-ins. 

(Aside: It would be great to allow themes to specify values other than `px` here, but I know that's all we allow for the font-size one currently). 

## Required

The main "requirement" for this to land is relatively simple: that it doesn't break many themes out of the box. 😄 But I'll go through that list above: 

> 1. Similar to wide/full images, we offer a theme opt-in for custom margins in general. If this is not enabled, these controls don't show up. That way there's basically no chance this will break any theme.

This is (probably?) not a requirement. I think we can eliminate the general theme-opt-in by adopting something like you mention above — the logic that disables the UI and impact of margins for full-width group blocks. This is somewhat unideal though, so we may need to add some sort of language in the UI to explain what's going on. 

Alternatively: the left/right margin is what causes the problems here, so another way to handle this would be to ignore that in the case of full blocks, and have them only use the top/bottom margin. Maybe by default, those are just ignored for full-width blocks? Again, this is not totally ideal, and would likely require some explainer text. 

> 2. We offer a theme opt-in for themes to provide their own custom sizes. This is the equivalent of the [custom font size](https://developer.wordpress.org/block-editor/developers/themes/theme-support/#block-font-sizes) options Gutenberg already provides. 

Assuming we figure that first part out, the theme-supplied custom sizes are something that _should_ launch with this feature. But I'm not totally sure I'd say it's a requirement. It definitely needs to happen, but maybe not in this exact PR. 

> 3. We update the default themes to support both of those.

Whatever we end up doing for the first two points, the default themes will need to need to be tested and updated to support this feature. That's definitely a requirement. 

> 4. We write up documentation for theme developers to adopt both. 

If this _does_ require any themes to update, then the documentation will obviously have to be updated alongside this PR. So I'd say that is a requirement. 

---

I _think_ that covers everything, but I'd love some feedback from other theme developers as well.

allancole commented:
> A base concern I have about this change is: How can we make this manageable for themes? On a basic level, we'll want themes to be able to specify what exactly these sizes mean. (In the near term at least). Similar to how themes can provide custom font sizes.

Spent a little time digesting this and I think I have a slightly different opinion on this. I feel like themes should probably to stay out of the way of these per-block style settings. If anything, themes should come with one standard default spacing value and then if a customer changes the spacing in the editor, the theme should just respect the override. 

> I'm relatively certain that the main point of difficulty from theme development is the fact that any margins that are added here will need to be overwritten by themes for any full-width innerblocks.

Yeah, this is a super tricky condition to account for. My feeling on this is that spacing controls should only have an _inward_ effect on blocks ([example](https://cloudup.com/cKqHBmYE1Op)).  Outer margins should not be effected so we can maintain the behavior in `.alignwide`, `.alignfull` and other utility classes. If users want more space _between_ two blocks vertically, the Spacer Block is the probably a better method, I think. In that way, this tool might only need to effect padding. 

> Similar to wide/full images, we offer a theme opt-in for custom padding in general. If this is not enabled, these controls don't show up. That way there's basically no chance this will break any theme.

If these are rolled out with utility classes, we might not need to worry about the theme breaking at all. I don’t see something like `.has-large-top-padding { padding-top: 24px }` breaking any theme. CoBlocks and Kioken do a similar thing that’s worth experimenting with if you haven’t already. Both rely on plugin-based utility classes that essentially can work with any theme that doesn’t entirely overwrite Gutenberg the styles. 

> We offer a theme opt-in for themes to provide their own custom sizes. This is the equivalent of the custom font size options Gutenberg already provides.
>> Assuming we figure that first part out, the theme-supplied custom sizes are something that should launch with this feature. But I'm not totally sure I'd say it's a requirement. It definitely needs to happen, but maybe not in this exact PR.

Yeah, I kind of feel the same about this. I’m not sure it _needs_ to happen just yet. Even if themes come with a predefined set of values, customers should still be able to choose their own value (just like font-sizes). Because of this, I think this could launch without theme specific values, as the arbitrary values will still be useful.

getdave commented:
> Good questions! I definitely don't think any of this is insurmountable! So that's the good news. 😄

That is good news 😄 

> We could theoretically make 1 & 2 variations of the same opt-in

_Custom_ spacing should be opt-in. However, I think most users would expect the ability to change padding/margin to be part of a default base experience. Therefore I'm not sure we should ask Themes to opt in to allowing the spacing controls to show up.

> It would be great to allow themes to specify values other than px here

That's probably possible. I'm not clear as to why font sizes only allows for `px`. Is there a reason?

> The main "requirement" for this to land is relatively simple: that it doesn't break many themes out of the box.

Agreed.

> I think we can eliminate the general theme-opt-in by adopting something like you mention above — the logic that disables the UI and impact of margins for full-width group blocks. This is somewhat unideal though, so we may need to add some sort of language in the UI to explain what's going on.

I think (as suggested above) this is a good approach. We might need some explanation text yes. 

> Assuming we figure that first part out, the theme-supplied custom sizes are something that should launch with this feature. But I'm not totally sure I'd say it's a requirement. 

I'd say it's a separate PR. We land with sensible defaults and then add another PR for custom spacing. Once that's in we integrate into the work in this PR.

> ...the default themes will need to need to be tested and updated to support this feature. T

Agreed. A requirement that is outside of this PR but will need to be completed before we hit "Merged" on this one. Who could help with that? Yourself? 

> ...the documentation will obviously have to be updated alongside this PR. So I'd say that is a requirement.

Yep a docs update is required.

Thanks so much for your feedback. Looks like @allancole has left more which I'll respond to now.

getdave commented:
@allancole Thanks for your feedback. Sorry it's taken so long to come back you!

> If anything, themes should come with one standard default spacing value and then if a customer changes the spacing in the editor, the theme should just respect the override.

I'd agree that once we have the ability to manually define values then the user selection within the Editor should overide any default of Theme-specific settings. However, this PR doesn't reach that stage. All we do is add classes to the DOM elements in question and the editor/theme can/could choose to apply whatever styling they like to these to deliver the correct spacing. This is why I've gone for arbitrary `small`, `medium`, `large`...etc. It doesn't bind us to explicit sizes and provides a good balance of flexibility whilst preserving simplicity. 

> My feeling on this is that spacing controls should only have an inward effect on blocks

I'm happy with that if that's what your experience as a Theme developer tells us. This PR just provides the controls. We can choose to make the control for `padding` or `margin` or both as required. If it's easier to ignore `margin` for now and that helps us land this PR then so much the better. Do you agree @kjellr ?

> If these are rolled out with utility classes, we might not need to worry about the theme breaking at all. I don’t see something like .has-large-top-padding { padding-top: 24px } breaking any theme.

They will if these styles are included in the editor default styles on a per Block basis. The parent PR has [an example of this](https://github.com/WordPress/gutenberg/pull/16730/files#diff-a1678e3d2dad77286726884a772e9940). Without providing defaults this functionality won't work unless the Theme supports it. As I indicated above in response to @kjellr I feel that users would expect this to work "out of the box" across all Themes. I'm very open to an alternative perspective on this however.

> Even if themes come with a predefined set of values, customers should still be able to choose their own value (just like font-sizes). Because of this, I think this could launch without theme-specific values, as the arbitrary values will still be useful.

I didn't entirely follow. To be clear I"m against introducing _manually_ customizable values in the Editor at this stage (ie: where you could tweak each value in `px` _directly_ in the Editor).

karmatosed commented:
I have to echo other sentiments here that how this plays out across the wide range of themes is worrisome. There is prior art in theme_support but it's what we decide on the default to be.

@kjellr has a strong point about the ideal state. I agree that the way full/wide images has been done stands us in a good place for this. I would add noting this new change in dev docs is key for release. It could fit nicely into WordPress 5.3.

My vote is for that as doesn't break anything out of the box.

I would +1 ignoring padding when the block is full width. It isn't ideal but it seems like the best default.

allancole commented:
> It doesn't bind us to explicit sizes and provides a good balance of flexibility whilst preserving simplicity.

Okay, that’s great to know. I’m probably thinking a bit too far ahead, but as long as there’s a good amount of flexibility, I’m on board :-) 

> This PR just provides the controls. We can choose to make the control for padding or margin or both as required. If it's easier to ignore margin for now and that helps us land this PR then so much the better. 

As I think on this more, I can definitely see situations where I would want to have both. Spacer block is good for adding _extra_ vertical space between blocks, but sometimes I want to _remove_ space and to do that, we’d need a feature like what you’ve suggested here. 

> I don’t see something like .has-large-top-padding { padding-top: 24px } breaking any theme.
>> They will if these styles are included in the editor default styles on a per Block basis. 

I’m sure I’m probably missing some context here, but do these styles _need_ to be added on a per-block basis? I’m imagining myself using this on any block where I might want to arbitrarily alter the spacing in some way. Seems like all blocks—nested or otherwise—could benefit from this kind of spacing flexibility so utility classes seems like an easier method to me. 

I’ll also say there’s quite a few plugins that offer this same functionality. The two I’m thinking of: CoBlocks and Kioken, both rely on utility classes and have (almost) zero conflicts with theme styles. That’s definitely influencing my expectation here, but its very possible I’m missing the context that makes per-block styles more appropriate. 

> To be clear I"m against introducing manually customizable values in the Editor at this stage (ie: where you could tweak each value in px directly in the Editor).

Sure, I’m okay with that not making it into this stage. 👍

dianeco commented:
As a webdesigner, I find that the Group block is greatly missing padding controls to build a website and I hope it will make it into WordPress 5.3
Simply adding a control to select padding size (none, small, medium, large...) would be very useful.
Using classes allow theme developers to change the values and make it responsive if they want.
Margin isn't necessary as we can use the Spacer block to create spacing. 

As mentioned in the Core Editor chat by yannicki, the current way to add padding inside a group is to insert spacers, which isn't ideal: "The new demo themes on wordpress.com (https://stratforddemo.wordpress.com/) are using 2 spacers (1 at the top, 1 at the bottom) to create padding inside each group because there are no other options available."

getdave commented:
Hey. I've been away for a while. I'm back now and looking to give this PR some attention.

getdave committed:
Update default value label

paaljoachim commented:
@getdave @youknowriad @jasmussen @kjellr 

What is still needed to get the most basic/first iteration/MVP merged? 

I do believe it will be very helpful for users to have additional ways to create padding/margins in their layouts. This feature can also be highlighted in the Twenty Twenty theme.

afercia commented:
From an accessibility perspective, the only issue I can see is about the label used on the _icon_.

The icon gets a `iconLabel` prop that is rendered as a `label` HTML attribute, which is invalid HTML:

<img width="693" alt="Screenshot 2019-09-13 at 10 34 54" src="https://user-images.githubusercontent.com/1682452/64850325-7c534500-d615-11e9-8027-88fb89751164.png">

The actual prop set on the `<Icon>` should be changed to `aria-label`.

However, this wouldn't have any effect because the Gutenberg icons are hidden from assistive technologies with `aria-hidden="true"`.  In this case, the text "All Devices" won't be announced by screen readers.

The icon should be only decorative with no aria-label. Proper labeling should be done via visible text.

To my understanding, this control is meant to add padding / margin settings to the group block, with the ability to specify this setting for different types of devices. If this is the case, I'd suggest to use `<fieldset>` and `<legend>` elements to group the settings based on devices. Pseudo code:
```
fieldset
    visible legend: All devices
    DimensionControl: Margin
    DimensionControl: Padding
/fieldset
fieldset
    visible legend: Small screen devices
    DimensionControl: Margin
    DimensionControl: Padding
/fieldset
etc.
```

Aside: I see there are a few things to address in this PR, for example: translatable strings, doc blocks, etc. Will comment inline.

afercia commented on packages/block-editor/src/components/dimension-control/README.md:
```diff
@@ -0,0 +1,138 @@
+			<DimensionControl
+				label={ __( 'Padding' ) }
+				icon={ 'desktop' }
+				iconLabel={ 'Desktop Devices' }
```
Please remove the iconLabel

getdave replied:
Understood. Thanks for feedback here. I've actioned this.

afercia commented on packages/block-editor/src/components/dimension-control/README.md:
````diff
@@ -0,0 +1,138 @@
+```
+
+_Note:_ it is recommend to partially apply the value of the Block attribute to be updated (eg: `paddingSize`, `marginSize`...etc) to your callback functions. This avoids the need to unecessarily couple the component to the Block attribute schema.
+
````
Please check typos, dots, periods, etc.

getdave replied:
👍

afercia commented on packages/block-editor/src/components/dimension-control/README.md:
````diff
@@ -0,0 +1,138 @@
+```
+
+By default a set of relative sizes (`small`, `medium`...etc) are provided. See `packages/block-editor/src/components/dimension-control/sizes.js`.
+
````
Please check commas, dots, periods.

getdave replied:
If you have noticed something specific it'd be super helpful if you could pinpoint it. I will check but if I missed it first time I could miss it again. Much appreciated.

afercia commented on packages/block-editor/src/components/dimension-control/README.md:
```diff
@@ -0,0 +1,138 @@
+* **Default:** `''`
+* **Required:** No
+
+A label for the icon (see above).
```
Please remove the iconLabel, see https://github.com/WordPress/gutenberg/pull/16791#issuecomment-531163401

afercia commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,101 @@
+import sizesTable, { findSizeBySlug } from './sizes';
+
+export function DimensionControl( props ) {
+	const { label, icon, iconLabel = 'all', sizes = sizesTable, currentSize, onSpacingChange, onReset, className = '' } = props;
```
Please remove iconLabel

afercia commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,101 @@
+	 * to a format suitable for use in the
+	 * <SelectControl /> options prop
+	 * @param  {Array} theSizes the array of sizes objects
+	 * @return {Array}       the array of options with default option prepended
```
Please adjust the formatting of docblocks to be consistent with Gutenberg practices (see also the previous docblocks)

getdave replied:
Honestly, I couldn't easily locate the best practices which says something in and of itself. I've reached out to Riad to ask whether there's tooling in place to auto-format these because relying on humans to do this sort of thing might not be a great way forward.

afercia replied:
Here's the WordPress ones:

JavaScript Documentation Standards
https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/

The Gutenberg ones slightly differ for minor details, not sure they're documented anywhere.

afercia commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,101 @@
+			{ icon && (
+				<Icon
+					icon={ icon }
+					label={ iconLabel || '' }
```
Please remove the `label` prop: icons should only be decorative, see https://github.com/WordPress/gutenberg/pull/16791#issuecomment-531163401

getdave replied:
Done.

afercia commented on packages/block-editor/src/components/dimension-control/sizes.js:
```diff
@@ -0,0 +1,54 @@
+ * all hardcoded `size` values are based on the value of
+ * the Sass variable `$block-padding` from
+ * `packages/block-editor/src/components/dimension-control/sizes.js`.
+ */
```
Please adjust this docblock to meet the Gutenberg practices.

afercia commented on packages/block-editor/src/components/dimension-control/sizes.js:
```diff
@@ -0,0 +1,54 @@
+ * @param  {Array} sizes containing objects for each size definition
+ * @param  {string} slug a string representation of the size (eg: `medium`)
+ * @return {Object}       the matching size definition
+ */
```
Please adjust this docblock to meet the Gutenberg practices.

afercia commented on packages/block-library/src/group/edit.js:
```diff
@@ -43,6 +48,13 @@ function GroupEdit( {
 				/>
+				<PanelBody title={ __( 'Spacing' ) }>
+					<DimensionControl
+						label={ 'Padding' }
```
This string is not translatable: should use `__()`

getdave replied:
Whilst you are entirely correct, this code was purely for testing purposes (I did note this in description but it was easy to miss). It has been removed as this is now close to merge.

afercia commented on packages/block-library/src/group/edit.js:
```diff
@@ -43,6 +48,13 @@ function GroupEdit( {
+					<DimensionControl
+						label={ 'Padding' }
+						icon={ 'desktop' }
+						iconLabel={ 'All Devices' }
```
This string is not translatable: should use `__()`

getdave replied:
Whilst you are entirely correct, this code was purely for testing purposes (I did note this in description but it was easy to miss). It has been removed as this is now close to merge.

getdave committed:
Removes unnecessary InstanceId HOC usage

Addresses https://github.com/WordPress/gutenberg/pull/16791#discussion_r323906696

getdave commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,101 @@
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+import { withInstanceId } from '@wordpress/compose';
```
Actually no we aren't. I've removed it.

getdave committed:
Remove unused abbreviation in size table

getdave committed:
Revert "Adds temp testing example usage of component to Group Block"

This reverts commit 6f9f3bfd2a7c1a08ecfab143384d414701f0c1e8.

getdave commented on packages/block-editor/src/components/dimension-control/sizes.js:
```diff
@@ -0,0 +1,54 @@
+export default [
+	{
+		name: __( 'None' ),
+		abbr: __( 'No' ),
```
Nope!

getdave committed:
Remove arbitrary size value from sizes list

This is not required as we cannot know how the dimensions component will be used. Therefore sticking with relative values via the slugs is safer. These can be mapped on a case by case basis as required.

getdave committed:
Remove icon label for a11y reasons

Addresses https://github.com/WordPress/gutenberg/pull/16791#discussion_r324103481

getdave committed:
Update component docs for consistency, spelling and grammar

getdave committed:
Tweak docblock formats

getdave commented:
@afercia Thanks for taking the time to review this. 

> To my understanding, this control is meant to add padding / margin settings to the group block, with the ability to specify this setting for different types of devices. 

It's confusing. As per the description, this is a sub PR of the main PR which adds the functionality you are describing. I was asked to create smaller PRs for these components because the main PR was getting unwieldy. It's been a source of confusion ever since!

> I'd suggest to use <fieldset> and <legend> elements to group the settings based on devices. 

This is great feedback. It's probably more relevant to either the parent PR or [the one that deals with a Responsive Control specifically](https://github.com/WordPress/gutenberg/pull/16790).

getdave committed:
Update test snapshots to match new default value

afercia commented:
@getdave thank you, will follow the parent issue.

getdave commented:
I think this is ready for a final review @youknowriad .

jorgefilipecosta commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,98 @@
+ */
+import sizesTable, { findSizeBySlug } from './sizes';
+
+export function DimensionControl( props ) {
```
Why not add this to the components folder?
It seems generic and has no dependencies at all on the block editor?

getdave replied:
Probably a good point. I'm trying to recall why I originally added this to the Block Editor components. Could be a hangover of having decoupled this component from a larger PR.

If you're recommending moving it I'm happy to go with that.

jorgefilipecosta replied:
I think the component is in the right place now 👍

jorgefilipecosta commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,98 @@
+import sizesTable, { findSizeBySlug } from './sizes';
+
+export function DimensionControl( props ) {
+	const { label, icon, sizes = sizesTable, currentSize, onSpacingChange, onReset, className = '' } = props;
```
I guess this may be used on other things besides spacing, should we call the prop onSpacingChange simply onChange?

getdave replied:
Fixed

jorgefilipecosta commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,98 @@
+import sizesTable, { findSizeBySlug } from './sizes';
+
+export function DimensionControl( props ) {
+	const { label, icon, sizes = sizesTable, currentSize, onSpacingChange, onReset, className = '' } = props;
```
Normally our components receive the current value property as simply value.
E.g: the ColorPalette receives a value prop and not a currentColor property.

Would it make sense to rename currentSize to value for consistency reasons?

getdave replied:
Fixed

jorgefilipecosta commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,98 @@
+import sizesTable, { findSizeBySlug } from './sizes';
+
+export function DimensionControl( props ) {
+	const { label, icon, sizes = sizesTable, currentSize, onSpacingChange, onReset, className = '' } = props;
```
Normally to reset the value in our components onChange is called with an undefined value. Would the same work here and remove onReset be an option?

getdave replied:
Good question. I'll take a look as I have a feeling there was a specific reason for it being implemented like it is.

getdave replied:
Ok done

getdave replied:
Nope. All fixed.

jorgefilipecosta commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,98 @@
+	 * @param  {string} val the DOMEvent event.target
+	 * @return {void}
+	 */
+	const onChangeSpacingSize = ( val ) => {
```
I guess we can wrap the function definitions with useCallback to avoid creating a new function on each rerender.

getdave replied:
Can I check why you are suggesting we use `useCallback`  here? My understanding that it was effectively to "memoize" computationally expensive functions and/or to ensure referential equality.

I assume you're coming at this with a perf lense? If so do you feel that `onChangeSpacingSize` is sufficiently intensive to warrant a `useCallback`?

If I'm missed something I'd be open to an explanation - I'm always looking to improve my understanding, especially in the field of hooks!

jorgefilipecosta replied:
Hi @getdave, 

> My understanding that it was effectively to "memoize" computationally expensive functions and/or to ensure referential equality.

We share the understanding. I'm suggesting use callback to ensure referential equality.
Without use callback, we will pass new onChange value to SelectControl on each render. This makes SelectControl need to rerender and makes React need to update the dom events to reference a new callback. All of this seems wasted work and use callback makes it simple to avoid this wasted work.

getdave replied:
🙇 Thanks. I needed to grok that. Much appreciated.

jorgefilipecosta reviewed (commented):
It seems the PR in its current form does not contain the test code on the group block.
So before testing I added the following test code:
https://github.com/WordPress/gutenberg/commit/bb4be23045506d2f2811d0e0ef4e06c9fe6e1062

The DimensionControl component seems very similar to SelectControl. It adds an icon as an option. Would it be possible to mix this component and select control by adding an icon prop in Select-control? Or instead, rename this component to something like IconSelect similar to the IconButton (which is a button with an icon)?

getdave commented:
@jorgefilipecosta Thanks for your feedback. I'll address the comments shortly.

> The DimensionControl component seems very similar to SelectControl

I understand where you're coming from. I'm also conscious that we don't want to increase the AP surface area for Gutenberg.

However, in this case I feel there's a strong argument for this being a bespoke control even if it is quite similar to the underlying `SelectControl` component. My primary rationale here is that we need to ensure we're providing strong UI/UX guidance for the pattern of a control for handling dimensions.

There has been a lot of back and forth on this PR (and others) discussing the most appropriate type of control to handle dimensions. Initially, for example, originally I used x4 buttons inside a `ButtonGroup` and this was amended to its current form of a `select` only after @mtias and others provided guidance that it would be more appropriate as a `select`.

Given that there are probably a number of ways you _could_ choose to implement a control to handle dimensions, it feels like we need to be providing a **single unified interface** for this pattern. 

By creating a bespoke `DimensionControl` component we're signalling to folks that this is a component which has been through a vetting process in terms of design and UX and is the most appropriate choice when implementing this type of control.

I'm open to being told I'm wrong, but I can't help feeling that if we don't give strong direction we'll merely end up with infinite micro variations on this type of control leaving the Gutenberg UX fragmented across the ecosystem.

getdave committed:
Update API from onSpacingChange to more agnostic onChange

Addresses https://github.com/WordPress/gutenberg/pull/16791#discussion_r331622801

getdave committed:
Update tests to cover onChange handler renamed

getdave committed:
Update currentSize prop to value for consistency with other components

getdave committed:
Removes onReset in favour of onChange with undefined for consistency

Adddresses https://github.com/WordPress/gutenberg/pull/16791#discussion_r331624272

getdave committed:
Move component to @wordpress/components package

getdave committed:
Remove invalid font sizes style import

Accidentally included from rebase.

getdave committed:
Deps update due to rebase

jorgefilipecosta commented on packages/components/src/dimension-control/index.js:
```diff
@@ -0,0 +1,92 @@
+export function DimensionControl( props ) {
+	const { label, icon, sizes = sizesTable, value, onChange, className = '' } = props;
+
+	/**
```
Spacing issue.

jorgefilipecosta commented on package-lock.json:
```diff
@@ -7636,12 +7636,7 @@
 			"version": "file:packages/env",
```
Should this package-lock changes be reverted? We are not adding a new package in this PR.

getdave replied:
I've reverted this.

jorgefilipecosta commented on packages/components/src/dimension-control/README.md:
```diff
@@ -0,0 +1,128 @@
+DimensionControl
+=============================
+
+`DimensionControl` is a component designed to provide a UI to control spacing and/or dimensions. It is intended for use within the editor `InspectorControls` sidebar.
```
I think we should not refer "It is intended for use within the editor `InspectorControls` sidebar" nothing in the component makes it less usable elsewhere e.g: in settings modal.

getdave replied:
Agreed.

jorgefilipecosta commented on packages/components/src/dimension-control/README.md:
```diff
@@ -0,0 +1,128 @@
+
+An optional dashicon to display before to the control label.
+
+### `onSpacingChange`
```
The docs don't reflect the new names of the props e.g: onSpacingChange is now onChange.

getdave replied:
Updated

jorgefilipecosta commented on packages/components/src/index.js:
```diff
@@ -12,6 +12,7 @@ export { default as ColorPalette } from './color-palette';
 export { default as ColorPicker } from './color-picker';
 export { default as Dashicon } from './dashicon';
 export { DateTimePicker, DatePicker, TimePicker } from './date-time';
+export { default as DimensionControl } from './dimension-control';
```
I guess it would be good to export this component as __experimentalDimensionControl, until we test it in practice. It makes it easier to do the required API changes in the future.

getdave replied:
Good plan. Done.

jorgefilipecosta reviewed (approved):
I left some comments (minor changes) that I think should be addressed before the merge but other than that the PR looks good to me from the code point of view 👍  Thank you for the iterations.
I'm cc(ing) @mapk, and @davewhitley, as they have done some work related to the components package and may have some thoughts on the new component.

getdave committed:
Remove unneeded doc blocks

getdave committed:
Remove usage suggestion which was not helpful

getdave committed:
Update readme docs to match current API

Addresses https://github.com/WordPress/gutenberg/pull/16791#discussion_r332692714

getdave committed:
Export as experimental component

Addresses https://github.com/WordPress/gutenberg/pull/16791#discussion_r332694561

getdave committed:
Revert "Deps update due to rebase"

This reverts commit 95d00f39010edfaac620980e0d0e7c1001a68c98.

Addresses https://github.com/WordPress/gutenberg/pull/16791#discussion_r332691520

youknowriad commented on packages/components/src/dimension-control/index.js:
```diff
@@ -0,0 +1,76 @@
+ */
+import sizesTable, { findSizeBySlug } from './sizes';
+
+export function DimensionControl( props ) {
```
I'm personally a bit hesitant on making this a public component in `@wordpress/components`.

The `@wordpress/components` package is basically the design system while this is more something for the editor, maybe something for the `block-editor` package?

What do you all think about moving this to the block-editor package or making it experimental?

youknowriad replied:
Nvm, I just saw that it's already experimental :) I wonder if we should move it to the block-editor and stabilize it once we use it a block.
